repair/repair.cc

```diff
@@ -1386,7 +1404,11 @@ static future<> do_repair_ranges(lw_shared_ptr<repair_info> ri) {
                 ri->ranges_index++;
                 rlogger.info("Repair {} out of {} ranges, id={}, shard={}, keyspace={}, table={}, range={}",
                     ri->ranges_index, ri->ranges.size(), ri->id, ri->shard, ri->keyspace, ri->table_names(), range);
-                return repair_range(*ri, range);
+                return repair_range(*ri, range).then([ri] {
+                    if (ri->reason == streaming::stream_reason::bootstrap) {
+                        _node_ops_metrics.bootstrap_finished_ranges++;
+                    }
+                });
```

Comment:
We'll need to undo on failure, in order to preserve consistency with total.

Reply:
On failure, the finished is not increased at all. We have nothing to undo.

Reply:
Whenever finished != total, it means that something is in progress. So we have to do something.

Reply:
Two metrics: finished and total can not express the information "failed".  User can know an operation is failed by other means, e.g., api. If the operation is failed, the user  know it is partially done.

Reply:
We will have to set the metrics to something. We can drop them both to zero after the operation completes (one way or the other).

Reply:
Setting to zero after the operation makes it is hard to use the metrics. It is hard to tell if the operation has finished or hasn't started yet.

Reply:
Well, we can't leave it at a random place. The graphs well look terrible.

Reply:
It is not a random place. It is a place where either it stopped progressing due to error or it is 100% in the case of successful operation.

Reply:
Those graphs that are stuck in some place just look bad (this example looks especially bad because I drew it)

![Drawing](https://user-images.githubusercontent.com/1017210/88029767-17c11a80-cb43-11ea-91c1-9493a3f68335.png)

It looks like a bug.

Reply:
LOL. 

You wanted this (dropping to zero). 

![2020-07-21_16-32](https://user-images.githubusercontent.com/320076/88031893-ced38b00-cb6f-11ea-88c7-49fe45fcf1c8.png)

Reply:
Dropping to zero has the issue the dashboard hasn't sampled the 100% progress while the progress dropped to zero already. We have false unfinished progress in the graph.

Reply:
Well, you're a better drawer than me. You're right about not sampling 100%.

Maybe @amnonh can add annotations about repair start and repair end/fail to the graph.